artiq/language/scan.py

```diff
@@ -59,52 +59,28 @@ def describe(self):
                 "repetitions": self.repetitions}
 
 
-class LinearScan(ScanObject):
-    """A scan object that yields a fixed number of evenly
-    spaced values in a range."""
-    def __init__(self, start, stop, npoints):
+class RangeScan(ScanObject):
+    """A scan object that yields a fixed number of evenly spaced values in a
+    range. If ``randomize`` is True the points are randomly ordered."""
+    def __init__(self, start, stop, npoints, randomize=False, seed=None):
         self.start = start
         self.stop = stop
         self.npoints = npoints
+        self.randomize = randomize
+        self.seed = seed
 
-    @portable
-    def _gen(self):
-        if self.npoints == 0:
-            return
-        if self.npoints == 1:
-            yield self.start
+        if npoints == 0:
+            self.sequence = []
+        if npoints == 1:
+            self.sequence = [self.start]
         else:
-            dx = (self.stop - self.start)/(self.npoints - 1)
-            for i in range(self.npoints):
-                yield i*dx + self.start
-
-    @portable
-    def __iter__(self):
-        return self._gen()
-
-    def __len__(self):
-        return self.npoints
+            dx = (stop - start)/(npoints - 1)
+            self.sequence = [i*dx + start for i in range(npoints)]
 
-    def describe(self):
-        return {"ty": "LinearScan",
-                "start": self.start, "stop": self.stop,
-                "npoints": self.npoints}
-
-
-class RandomScan(ScanObject):
-    """A scan object that yields a fixed number of randomly ordered evenly
-    spaced values in a range."""
-    def __init__(self, start, stop, npoints, seed=None):
-        self.start = start
-        self.stop = stop
-        self.npoints = npoints
-        self.seed = seed
-        self.sequence = list(LinearScan(start, stop, npoints))
-        if seed is None:
-            rf = random.random
-        else:
-            rf = Random(seed).random
-        random.shuffle(self.sequence, rf)
+        if randomize:
+            if seed is not None:
+                random.seed(seed)
```

Comment:
Even if this has no practical consequences here, I'd like to avoid touching global state when that can be avoided, which is why I was using ``random.Random`` in the original code. Your seeding code would break if there are two threads with different seeds executing this at the same time.